### PR TITLE
Add support for building on and targeting Mono

### PR DESCRIPTION
**Pull request by radical**
To build for mono:
    `$ ./build.sh -host mono`

This will download a bootstrap msbuild/mono and use the system (in $PATH) mono to build.
You can use `-useSystemMSBuild` argument to skip the bootstrap msbuild/mono and just
use the system msbuild.

### CODE REVIEW

**Comment by rainersigwald** [comment]
@radical I think mono CI jobs are disabled at the moment. Let me try a couple of things.

**Comment by rainersigwald** [comment]
@dotnet-bot test ci please

**Comment by radical** [comment]
@dotnet-bot test ci please

**Comment by radical** [comment]
@dotnet-bot test ci please

**Comment by radical** [comment]
@dotnet-bot test mono

**Comment by radical** [comment]
@dotnet-bot test mono

File: build/build.sh [hunk 1]
```diff
@@ -27,6 +27,7 @@ dotnetCoreSdkDir=""
 function Help() {
   echo "Common settings:"
   echo "  -configuration <value>  Build configuration Debug, Release"
+  echo "  -host <value>           core (default), mono"
```
Comment (cdmihai) [inline build/build.sh]: nit: in the .ps1 we call it `hostType`. It would be nice if the parameter names match.

File: build/build.sh [hunk 2]
```diff
@@ -176,6 +190,20 @@ function GetLogCmd {
...
 
+function downloadMSBuildForMono {
+  if [[ -z $useSystemMSBuild && ! -e "$MONO_MSBUILD_DIR/MSBuild.dll" ]]
+  then
+    echo "** Downloading MSBUILD from $MSBUILD_DOWNLOAD_URL"
+    curl -sL -o "$MSBUILD_ZIP" "$MSBUILD_DOWNLOAD_URL"
+
+    unzip -q "$MSBUILD_ZIP" -d "$ArtifactsDir"
+    # rename just to make it obvious when reading logs!
+    mv $ArtifactsDir/msbuild $ArtifactsDir/mono-msbuild
```
Comment (cdmihai) [inline build/build.sh]: MONO_MSBUILD_DIR points to this path right?

File: build/build.sh [hunk 3]
```diff
@@ -279,9 +307,13 @@ function ErrorHostType {
 }
 
 function Build {
-  InstallDotNetCli
+  if [ "$host" = "core" ]; then
+    InstallDotNetCli
+    echo "Using dotnet from: $DOTNET_INSTALL_DIR"
+  else
+    CreateDirectory $ArtifactsDir
```
Comment (cdmihai) [inline build/build.sh]: Unconditionally create this dir

**Comment by cdmihai** [comment]
>     eval "$(QQ $DOTNET_HOST_PATH) nuget locals all --clear"

How will this work without dotnet?

---
Refers to: build/build.sh:320 in 3cd9063. [](commit_id = 3cd90634f1d4026c3577af42c0a95ec374cc2a85, deletion_comment = False)

File: build/build.sh [hunk 4]
```diff
@@ -325,11 +367,18 @@ function Build {
...
+      msbuildHost="mono"
+
+      properties="$properties /p:MSBuildExtensionsPath=$bootstrapRoot/net461/MSBuild/"
     else
       ErrorHostType
     fi
 
-    export ArtifactsDir="$ArtifactsDir/2"
+    # forcing the slashes here or they get normalized and lost later
+    export ArtifactsDir="$ArtifactsDir\\2\\"
```
Comment (cdmihai) [inline build/build.sh]: why backslashes?

File: src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs [hunk 5]
```diff
@@ -6473,26 +6473,6 @@ public void SubsetListFinderEmptySubsetToSearchFor()
             Assert.Equal(0, returnArray.Length); // "Expected the array returned to be 0 length"
         }
 
-
-        /// <summary>
-        /// Verify that the method will not crash if there are empty string array elements, and that when we call the
-        /// method twice with the same set of SubsetToSearchFor and TargetFrameworkDirectory that we get the exact same array back.
-        /// </summary>
-        [Fact]
-        public void SubsetListFinderVerifyEmptyInSubsetsToSearchForAndCaching()
```
Comment (cdmihai) [inline src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs]: Why did you delete this test?

Comment (radical) [reply src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs]: I explained it in the commit message:

```
[test] Remove an unnecessary test  …
RAR_Tests.SubsetListFinderVerifyEmptyInSubsetsToSearchForAndCaching

This is essentially checking if the the results of two different queries
are two *different* empty array objects. This is an implementation
detail and should not be relied upon. It is a legal optimization to
return the same empty array object, which is what List.ToArray does on
netcore and mono (on macOS). Full framework returns the same object, so
this test passed on .net/windows.
```

File: build/build.sh (hunk 4 above)
Comment (radical) [reply build/build.sh]: The `2Debug` directory names seemed like a mistake. This change would make the directories `artifacts/2/Debug-Mono/...`. But if that is intentional, then we can skip this change.

File: build/build.sh (hunk 2 above)
Comment (radical) [reply build/build.sh]: That's correct. I will update the line to use that.

File: src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs (hunk 5 above)
Comment (radical) [reply src/Tasks.UnitTests/AssemblyDependency/Miscellaneous.cs]: Maybe the test can be reduced to just `Verify that the method will not crash if there are empty string array elements,` ?

**Comment by radical** [comment]
Updated.

**Comment by radical** [comment]
Looking at the windows/full failure.

**Comment by cdmihai** [comment]
@dotnet-bot test ci please

**Comment by cdmihai** [comment]
@radical FYI, the RHEL job is broken, don't spend time on it.

**Comment by radical** [comment]
I'm not sure what is causing this. windows:

```
11:21:02 D:\j\w\innerloop_Win---c5dcd6be\src\Package\MSBuild.VSSetup\files.swr(22,2): error SWIX1108: Could not find file at path: MSBuildTaskHost.exe. [C:\Users\runner\.nuget\packages\roslyntools.repotoolset\1.0.0-beta2-62901-01\tools\VisualStudio.SetupPackage.swixproj]
11:21:02 D:\j\w\innerloop_Win---c5dcd6be\src\Package\MSBuild.VSSetup\files.swr(23,2): error SWIX1108: Could not find file at path: MSBuildTaskHost.exe.config. [C:\Users\runner\.nuget\packages\roslyntools.repotoolset\1.0.0-beta2-62901-01\tools\VisualStudio.SetupPackage.swixproj]
```

`MSBuildTaskHost` project should be importing the `ProducesNoOutput.Settings.props` only for:
```
-  <Import Project="$(RepoRoot)build\ProducesNoOutput.Settings.props" Condition="'$(MSBuildRuntimeType)' == 'Core' or '$(MachineIndependentBuild)' == 'true'" />
+  <Import Project="$(RepoRoot)build\ProducesNoOutput.Settings.props" Condition="'$(MSBuildRuntimeType)' == 'Core' or '$(MachineIndependentBuild)' == 'true' or '$(MonoBuild)' == 'true'" />
```

The only change here is `MonoBuild==true`, which shouldn't be the case here!

**Comment by radical** [comment]
oh, wait.. it passed? Did my earlier push not trigger the rebuild here?

**Comment by radical** [comment]
re:RHEL, ah, thanks!

**Comment by radical** [comment]
I'll wait to see a green mono build show up here, then I'll update the front page! Once this is merged, if I say `@dotnet-bot test ci please` on this PR itself then would the new build/lane show up?
And thanks for the review, @cdmihai and @rainersigwald  :)

**Comment by cdmihai** [comment]
Unfortunately the jobs won't appear until you commit to master. To get around that, you can manually run the test jobs generated by @dotnet-bot test ci please: https://ci2.dot.net/job/Microsoft_msbuild/job/master/job/GenPRTest/

I manually ran the OSX one, and it looks like it still downloads .net core: https://ci2.dot.net/job/Microsoft_msbuild/job/master/job/GenPRTest/job/innerloop_OSX10.13_MonoTest_prtest/1/console
The arguments are `Executed from ./build/cibuild.sh with arguments: -build -ci -prepareMachine -host mono`, so it looks like you have to update the groovy script with the new `-hostType` :)

**Comment by radical** [comment]
Since you have already reviewed the PR, I added the changes as new commits. Besides the `hostType` change, we have the fun:

```
[build] Correctly handle nuget cache clearing in case of mono  …
In case of mono, `$(DOTNET_HOST_PATH)` is not set and `$(QQ ..)` ends up
giving us `""` and so the eval call becomes:

    `eval("" nuget ...)`

and that fails with command not found!

Also, the nuget command for dothet and mono seems to differ slightly:

    `--clear` vs `-clear`(mono)
```

**Comment by radical** [comment]
@dotnet-bot test ci please

**Comment by radical** [comment]
re:`Is the important thing here the slash direction, or the addition of a trailing path separator? It seems like "$ArtifactsDir/2/" might be a better fix.`

So, I had to do this because the slashes were getting "normalized" and lost. I don't recall exactly where this was happening or what was doing this. I could dig more, if you want.

File: src/MSBuild.Bootstrap/MSBuild.Bootstrap.csproj [hunk 6]
```diff
@@ -21,7 +28,21 @@
     <Content Include="$(RepoRoot)artifacts\.dotnet\$(DotNetCliVersion)\sdk\$(DotNetCliVersion)\Microsoft.NETCoreSdk.BundledVersions.props" CopyToOutputDirectory="PreserveNewest" />
   </ItemGroup>
 
-  <ItemGroup>
+  <ItemGroup Condition="'$(MonoBuild)' == 'true'">
+    <!-- Include the bundled SDKs -->
+    <Content Include="$(RepoRoot)sdks\**\*" LinkBase="Sdks" CopyToOutputDirectory="PreserveNewest" />
```
[the remark this answers is not in the document]
Comment (radical) [reply src/MSBuild.Bootstrap/MSBuild.Bootstrap.csproj]: oh, and probably add Exists check!

**Comment by radical** [comment]
Updated.

**Comment by radical** [comment]
That `SkipTArgetFramework*` change required this:

```
-    <XunitOptions Condition="'$(MonoBuild)' == 'true'">$(XunitOptions) -notrait category=non-mono-tests</XunitOptions>
+    <XunitOptions Condition="'$(MonoBuild)' == 'true'">$(XunitOptions) -notrait category=non-mono-tests -notrait category=nonmonotests</XunitOptions>
```
Added as a new commit.
